### PR TITLE
Fixing flakes caused by petset tests.

### LINKED ISSUE #33904: [k8s.io] PetSet [Slow] [Feature:PetSet] [k8s.io] Deploy clustered applications [Slow] [Feature:PetSet] should creating a working redis cluster [Feature:PetSet] {Kubernetes e2e suite}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gci-gce-petset/540/

Failed: [k8s.io] PetSet [Slow] [Feature:PetSet] [k8s.io] Deploy clustered applications [Slow] [Feature:PetSet] should creating a working redis cluster [Feature:PetSet] {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/petset.go:223
Oct  3 02:45:20.585: Read unexpected value , expected bar under key foo
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/petset.go:221
```

### PR DESCRIPTION
**Pull request by foxish**
**What this PR does / why we need it**: We should wait till the key/value has been replicated to database slaves before trying to read from them.

**Which issue this PR fixes** Fixes https://github.com/kubernetes/kubernetes/issues/33904, https://github.com/kubernetes/kubernetes/issues/33895

``` release-note
Fixed flakes caused by petset tests.
```

cc @kubernetes/sig-apps

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35158)

### CODE REVIEW

File: test/e2e/petset.go [hunk 1]
```diff
@@ -216,6 +216,10 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
 			By("Creating foo:bar in member with index 0")
 			pet.write(0, map[string]string{"foo": "bar"})
 
+			// Wait for the write to complete to all slaves with a timeout of 100s (specified in ms).
+			By("waiting for all slaves to sync")
+			pet.waitForReplication(0, ps.Spec.Replicas - 1, 100000)
```
Comment (bprashanth) [inline test/e2e/petset.go]: oh, thought we were just polling on retry on the read. Please do that instead.

Comment (bprashanth) [inline test/e2e/petset.go]: for the other pets too, if you can

Comment (foxish) [inline test/e2e/petset.go]: I see, we don't want application-specific code here? Then I guess I could make a polling mechanism that would work for all of them.

Comment (bprashanth) [inline test/e2e/petset.go]: No i mean polling on kubectl exec is good for reasons other than replication lag, and would deal with the same issue. Existing pets should be using a common interface in this e2e (`petTester`), im asking that you change the read methods on that interface.

**Comment by foxish** [comment]
@bprashanth @janetkuo PTAL

File: test/e2e/petset.go [hunk 2]
```diff
@@ -804,6 +804,18 @@ func ExpectNoError(err error) {
 	Expect(err).NotTo(HaveOccurred())
 }
 
+func pollReadWithTimeout(pet petTester, petNumber int, key, expectedVal string) error {
+	return wait.PollImmediate(time.Second, timeout, func() (bool, error) {
```
Comment (bprashanth) [inline test/e2e/petset.go]: what is timeout? if it's a global we should rename it to something descriptive

File: test/e2e/petset.go [hunk 3]
```diff
@@ -223,8 +223,8 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
 			}
 
 			By("Reading value under foo from member with index 2")
-			if v := pet.read(2, "foo"); v != "bar" {
-				framework.Failf("Read unexpected value %v, expected bar under key foo", v)
+			if err := pollReadWithTimeout(pet, 2, "foo", "bar"); err != nil {
```
Comment (bprashanth) [inline test/e2e/petset.go]: this error is useless, it's the infamous "timed out waiting for condition". Maybe pipe the actual error reported from db or something? Or say timed out trying to read back value foo from replicated mysql slave...

File: test/e2e/petset.go (hunk 2 above)
Comment (foxish) [inline test/e2e/petset.go]: Fixed.

File: test/e2e/petset.go (hunk 3 above)
Comment (foxish) [inline test/e2e/petset.go]: Fixed by changing `pollReadWithTimeout` to return easier to read errors.

**Review by bprashanth — APPROVED** [review]

**Comment by bprashanth** [comment]
Thanks lgtm, you're failing go-fmt. Feel free to re-apply on fix.

**Comment by mtaufen** [comment]
sweeet

**Comment by chrislovecnm** [comment]
Do I need to look at the C* PetSet tests as well?? Will make a note to review this PR

**Comment by foxish** [comment]
@jessfraz This fixes https://github.com/kubernetes/kubernetes/issues/35589

**Comment by jessfraz** [comment]
oh awesome thanks!
I will cherry pick

**Comment by jessfraz** [comment]
cherrypicked in #35653

**Pull request merged**